### PR TITLE
ENG-11036: voltadmin save with no arguments

### PR DESCRIPTION
**Pull request by akhanzode**

### CODE REVIEW

File: lib/python/voltcli/voltadmin.d/save.py [hunk 1]
```diff
@@ -32,24 +32,41 @@
...
+                              default = None),
+        VOLT.StringOption('-n', '--nonce', 'nonce', 'the unique snapshot identifier (nonce)', default = None),
+        VOLT.StringOption('-d', '--directory', 'directory', 'the snapshot server directory', default = None)
     )
 )
 def save(runner):
-    uri = 'file://%s' % urllib.quote(runner.opts.directory)
-    nonce = runner.opts.nonce.replace('"', '\\"')
+    uri = None
+    if runner.opts.directory != None:
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/save.py]: `is None`

File: lib/python/voltcli/voltadmin.d/save.py [hunk 2]
```diff
@@ -32,24 +32,41 @@
...
+        uri = 'file://%s' % urllib.quote(runner.opts.directory)
+    nonce = None
+    if runner.opts.nonce != None:
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/save.py]: `is None`

File: src/frontend/org/voltdb/SnapshotInitiationInfo.java [hunk 3]
```diff
@@ -65,20 +66,23 @@ public SnapshotInitiationInfo(Object[] params) throws Exception
...
-        if (params.length == 3) {
-            parseLegacyParams(params);
-        }
-        else if (params.length == 1) {
-            parseJsonParams(params);
-        }
-        else {
-            throw new Exception("@SnapshotSave requires 3 parameters " +
-                    "(Path, nonce, and blocking) or alternatively a single JSON blob. ");
+        boolean checkNonceValidity = true;
```
Comment (vtkstef) [inline src/frontend/org/voltdb/SnapshotInitiationInfo.java]: where is this set?

File: src/frontend/org/voltdb/SnapshotInitiationInfo.java [hunk 4]
```diff
@@ -65,20 +66,23 @@ public SnapshotInitiationInfo(Object[] params) throws Exception
...
+                m_nonce = MAGIC_NONCE_PREFIX + System.currentTimeMillis();
+                m_stype = SnapshotPathType.SNAP_AUTO;
+                m_path = VoltDB.instance().getSnapshotPath();
+                break;
         }
 
-        if (m_nonce != null && (m_nonce.contains("-") || m_nonce.contains(","))) {
-            throw new Exception("Provided nonce " + m_nonce + " contains a prohibited character (- or ,)");
+        if (checkNonceValidity && m_nonce != null && (m_nonce.contains("-") || m_nonce.contains(",")) && m_nonce.startsWith(MAGIC_NONCE_PREFIX)) {
+            throw new Exception("Provided nonce " + m_nonce + " contains a prohibited character (- or ,) or starts with " + MAGIC_NONCE_PREFIX);
```
Comment (vtkstef) [inline src/frontend/org/voltdb/SnapshotInitiationInfo.java]: maybe IllegalArgumentException ?

File: src/frontend/org/voltdb/SnapshotInitiationInfo.java [hunk 5]
```diff
@@ -168,9 +172,14 @@ private void parseJsonParams(Object[] params) throws Exception
         }
 
         m_stype = SnapshotPathType.valueOf(jsObj.optString(SnapshotUtil.JSON_PATH_TYPE, SnapshotPathType.SNAP_PATH.toString()));
-        m_path = jsObj.getString(SnapshotUtil.JSON_URIPATH);
-        if (m_path.isEmpty()) {
-            throw new Exception("uripath cannot be empty");
+        if (jsObj.has(SnapshotUtil.JSON_URIPATH)) {
+            m_path = jsObj.getString(SnapshotUtil.JSON_URIPATH);
```
Comment (vtkstef) [inline src/frontend/org/voltdb/SnapshotInitiationInfo.java]: you can turn a `has` and `get` into one `jsObj.optString(SnapshotUtil.JSON_URIPATH,"")`

File: src/frontend/org/voltdb/SnapshotInitiationInfo.java [hunk 6]
```diff
@@ -182,10 +191,13 @@ private void parseJsonParams(Object[] params) throws Exception
                     " if this is a file path then you must prepend file://");
         }
         m_path = pathURI.getPath();
-
-        m_nonce = jsObj.getString(SnapshotUtil.JSON_NONCE);
-        if (m_nonce.isEmpty()) {
-            throw new Exception("nonce cannot be empty");
+        if (jsObj.has(SnapshotUtil.JSON_NONCE)) {
```
Comment (vtkstef) [inline src/frontend/org/voltdb/SnapshotInitiationInfo.java]: same here

File: src/frontend/org/voltdb/SnapshotInitiationInfo.java (hunk 4 above)
Comment (akhanzode) [inline src/frontend/org/voltdb/SnapshotInitiationInfo.java]: Sure left it to what was before.

File: lib/python/voltcli/voltadmin.d/save.py [hunk 7]
```diff
@@ -34,22 +34,52 @@
...
-    uri = 'file://%s' % urllib.quote(runner.opts.directory)
-    nonce = runner.opts.nonce.replace('"', '\\"')
+    uri = None
+    dir_specified = False
+    if runner.opts.directory != None:
+        uri = 'file://%s' % urllib.quote(runner.opts.directory)
+        dir_specified = True
+
+    nonce = None
+    if runner.opts.nonce != None:
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/save.py]: is None

File: lib/python/voltcli/voltadmin.d/save.py [hunk 8]
```diff
@@ -34,22 +34,52 @@
...
+
     if runner.opts.blocking:
         blocking = 'true'
     else:
         blocking = 'false'
-    raw_json_opts = ['uripath:"%s"' % (uri),
-                     'nonce:"%s"' % (nonce),
-                     'block:%s' % (blocking),
-                     'format:"%s"' % (runner.opts.format)]
+    if uri != None:
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/save.py]: is None

File: lib/python/voltcli/voltadmin.d/save.py [hunk 9]
```diff
@@ -34,22 +34,52 @@
...
+        if nonce != None:
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/save.py]: is None

File: lib/python/voltcli/voltadmin.d/save.py [hunk 10]
```diff
@@ -34,22 +34,52 @@
...
+            raw_json_opts = ['uripath:"%s"' % (uri),
+                             'nonce:"%s"' % (nonce),
+                             'block:%s' % (blocking),
+                             'format:"%s"' % (runner.opts.format)]
+        else:
+            raw_json_opts = ['uripath:"%s"' % (uri),
+                             'block:%s' % (blocking),
+                             'format:"%s"' % (runner.opts.format)]
+    else:
+        if nonce != None:
```
Comment (vtkstef) [inline lib/python/voltcli/voltadmin.d/save.py]: is None

File: lib/python/voltcli/voltadmin.d/save.py (hunk 7 above)
Comment (akhanzode) [inline lib/python/voltcli/voltadmin.d/save.py]: you mean is not None as its checking for !None . Same for all other comments.

**Pull request merged**